File: python3/july/day_24_All Paths From Source to Target.py
class Solution:
        
    def dfs(self, s):
        if(s == len(self.graph) - 1):
            self.path.append(len(self.graph)-1)
            self.res.append(self.path[::])
            self.path.pop()
            return;
        
        self.path.append(s)
        for i in range(len(self.graph[s])):
            self.dfs(self.graph[s][i])
        self.path.pop()

    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        self.res = []
        self.graph = graph
        self.path = []
        self.dfs(0)
        return self.res

File: cpp/july/day_24_All Paths From Source to Target.cpp
class Solution {
public:
    void dfs(vector<vector<int>> &graph,vector<vector<int>> &res, int s, vector<int> &path){
        if(s == graph.size() - 1){
            path.push_back(graph.size()-1);
            res.push_back(path);
            path.pop_back();
            return;
        }
        path.push_back(s);
        for(int i = 0; i < graph[s].size(); i++){
            dfs(graph, res, graph[s][i], path);
        }
        path.pop_back();
    }
    
    
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        vector<vector<int>> res;
        vector<int> path;
        dfs(graph, res, 0, path);
        return res;
    }
};
